04.ls/ls.rb

```diff
@@ -2,28 +2,129 @@
 # frozen_string_literal: true
 
 require 'optparse'
+require 'etc'
 
 MAX_COL_COUNT = 3
 SPACE_WIDTH = 2
 
+FILE_TYPE_LIST = { '01' => 'p', '02' => 'c', '04' => 'd', '06' => 'b', '10' => '-', '12' => 'l', '14' => 's' }.freeze
+
+def main
+  options, directory_paths = parse_options(ARGV)
+  # directory_pathsには複数のpathを指定することは許容しているが、現時点でファイル名を表示するのは1番目に指定したディレクトリのみにしている。
+  directory_path = directory_paths[0] || './'
+  file_names = fetch_file_names(directory_path, options)
+  output(file_names, directory_path, options)
+end
+
 def parse_options(argv)
   opt = OptionParser.new
   options = {}
   opt.on('-a') { |v| options[:a] = v }
   opt.on('-r') { |v| options[:r] = v }
+  opt.on('-l') { |v| options[:l] = v }
   directory_paths = opt.parse(argv)
   [options, directory_paths]
 end
 
 def fetch_file_names(target_directory_path, options)
   dotmatch_flag = options[:a] ? File::FNM_DOTMATCH : 0
   file_names = Dir.glob('*', dotmatch_flag, base: target_directory_path)
+
   options[:r] ? file_names.reverse : file_names
 end
 
-def output(file_names)
+def output(file_names, target_directory_path, options)
   return if file_names.empty?
 
+  if options[:l]
+    output_long_listing_format(file_names, target_directory_path)
+  else
+    output_default_format(file_names)
+  end
+end
+
+def output_long_listing_format(file_names, target_directory_path)
+  puts "total #{calc_block_count_total(file_names, target_directory_path)}"
+
+  details_by_file_name = get_details_by_file_name(file_names, target_directory_path)
+
+  file_names.each do |file_name|
+    details = details_by_file_name[file_name]
+    details_output_order = %i[stat_mode nlink username groupname size ctime]
+
+    details_output_order.each do |key|
+      width = get_max_width_by_detail(details_by_file_name, key)
+      if /^\d+$/.match?(details[key])
+        print details[key].rjust(width)
+      else
+        print details[key].ljust(width)
+      end
+      print ' '
+    end
+    puts file_name
+  end
+end
+
+def calc_block_count_total(file_names, target_directory_path)
+  block_count_total = 0
+  file_names.each do |file_name|
+    block_count_total += File.stat("#{target_directory_path}/#{file_name}").blocks
+  end
+  block_count_total
+end
+
+def get_details_by_file_name(file_names, target_directory_path)
+  details_by_file_name = {}
+  file_names.each do |file_name|
+    stat = File.stat("#{target_directory_path}/#{file_name}")
+    details = convert_stat_to_details(stat)
+    details_by_file_name[file_name] = details
+  end
+  details_by_file_name
+end
+
+def convert_stat_to_details(stat)
+  details = {}
+  details[:stat_mode] = convert_stat_mode_to_str(stat.mode)
```

Comment:
```suggestion
  {
    stat_mode: convert_stat_mode_to_str(stat.mode),
```

好みの問題ですが、ハッシュリテラル一発で書いてしまう手もあります。

Reply:
その書き方が頭から抜けていました。
details[:key]と何回も書くのを避けたいので、提案どおり修正させて頂きました。